[ObsUx] Add feedback form to apm

jennypavlova opened the pull request:
Closes #172506

## Summary

This PR adds a button to all APM pages navigating to a feedback form. The feedback button component exists in infra so in this PR the component is changed to meet the new requirements and moved to `observability_shared`. Now the feedback button component supports also prefilling the sanitized path (`entry.1876422621`). The requirements picked for the sanitized path are described in the issue - it should be for example `/app/apm/{page_name}` if the user is on the page with the exact path matching  `/app/apm/{page_name}` and `/app/apm/{page_name}*` for all subpages (for example `/app/apm/{page_name}/something/else`) - different test scenarios can be found in the unit test: [get_path_for_feedback.test.ts](https://github.com/elastic/kibana/compare/main...jennypavlova:kibana:172506-obsux-add-feedback-form-to-apm?expand=1#diff-6396807e61353509c44fa38488dfb741549e60f25126024b92596f6b7ac933b8)

## Testing
- Go to APM
- All APM pages should have a yellow button with the text: `Tell us what you think!`
- Hover on the button to see the prefilled properties and check if every property is prefilled
<img width="1920" alt="image" src="https://github.com/elastic/kibana/assets/14139027/b57adf04-4e7b-42bb-827f-db554dc4a842">

- Click on the form and check if the form is prefilled - The questions that should be prefilled in APM
   - Where is your Elastic cluster deployed? (same as infra)
      - By default, the pre-selected option will be on-prem
      - To test the case with the cloud option preselected add `xpack.cloud.id: 'elastic_kibana_dev'` to your `kibana.dev.yaml`
      - To test the serverless option run Kibana in serverless mode (observability config)
   - What version of Elastic are you using? (same as infra)
   - Where in the UI are you?
- After checking the APM go to the infra pages and check the feedback button/form



https://github.com/elastic/kibana/assets/14139027/52cc886f-95a9-4099-b047-93fc64036572

jennypavlova committed:
Move FeatureFeedbackButton to obs shared to make it sharable between infra and APM

jennypavlova committed:
[WIP] APM feedback link

jennypavlova committed:
Add kibana version and cluster type to APM feedback link

jennypavlova committed:
Add sanitized path to the feedback url

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

jennypavlova commented:
/ci

jennypavlova committed:
Merge branch 'main' into 172506-obsux-add-feedback-form-to-apm

jennypavlova commented:
/ci

jennypavlova committed:
Update translation key

jennypavlova committed:
Merge branch '172506-obsux-add-feedback-form-to-apm' of https://github.com/jennypavlova/kibana into 172506-obsux-add-feedback-form-to-apm

jennypavlova committed:
Merge branch 'main' into 172506-obsux-add-feedback-form-to-apm

jennypavlova commented:
/ci

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

crespocarlos commented on x-pack/plugins/infra/public/pages/metrics/hosts/index.tsx:
```diff
@@ -29,6 +29,9 @@ const HOSTS_FEEDBACK_LINK =
 export const HostsPage = () => {
   const { isLoading, loadSourceFailureMessage, loadSource, source } = useSourceContext();
+  const {
+    services: { kibanaVersion, isCloudEnv, isServerlessEnv },
```
Just realized that `kibanaVersion`,`isCloudEnv`,  and `isServerlessEnv` are under `services`. Should we move them out of the `services` object? They are just flags and not actual services

jennypavlova replied:
Do we want to have a separate object [here](https://github.com/elastic/kibana/blob/main/src/plugins/kibana_react/public/context/context.tsx/#L17) in the context for those values? Should we call it kibanaEnvironment or kibanaSettings maybe? (those flags are about the kibana deployment/version so not sure what the best name is 🤔 wdyt?)

jennypavlova replied:
As discussed I picked `kibanaEnvironment` and added the changes [here](https://github.com/elastic/kibana/pull/173758/commits/aaadb14d60cd9da76ea4a523148061201324a36e#diff-77464f5aecf6f5ea2dbf4e6e65b8edd91a35b6341963427639cc81a77ae27de0)

crespocarlos commented on x-pack/plugins/apm/public/utils/get_path_for_feedback.ts:
```diff
@@ -0,0 +1,26 @@
+};
+
+export const getPathForFeedback = (path: string) => {
+  const pathStartingFromApp = shortenPath(path, '/app');
```
Is there a safer way to get this `/app` ? I tried to find some constant or service that would provide this base path, but couldn't find any

jennypavlova replied:
I couldn't find it as a constant, I see it is used as a string in other places. I defined a constant here so we can reuse it in case we need to have it somewhere else

crespocarlos reviewed (commented):
Looks good 👏 ! The form is pre-filled as expected. Didn't test all possible combinations though. 

Left a few minor coments.

jennypavlova commented on x-pack/plugins/apm/public/plugin.ts:
```diff
@@ -137,6 +139,9 @@ export interface ApmPluginStartDeps {
   observabilityAIAssistant: ObservabilityAIAssistantPluginStart;
   dashboard: DashboardStart;
   metricsDataAccess: MetricsDataPluginStart;
+  kibanaVersion?: string;
+  isCloudEnv: boolean;
+  isServerlessEnv: boolean;
```
They are inside `ApmPluginStartDeps` only `telemetry` is in `ApmServices`. Indeed, they are not service, where do you think we should put them?

crespocarlos replied:
Could we pass them directly to `renderApp`?

jennypavlova committed:
CR Changes

jennypavlova committed:
Merge branch '172506-obsux-add-feedback-form-to-apm' of https://github.com/jennypavlova/kibana into 172506-obsux-add-feedback-form-to-apm

jennypavlova commented on x-pack/plugins/apm/public/utils/get_path_for_feedback.ts:
```diff
@@ -0,0 +1,26 @@
+  const pathParts = pathStartingFromApp.split('/');
+  const constructPath = `/${[pathParts[1], pathParts[2], pathParts[3]].join(
+    '/'
+  )}`;
```
Changed, thanks ✅

jennypavlova committed:
CR change: extract to kibanaEnvironment

vadimkibana reviewed (commented):
In the long term we are planning to remove the `KibanaReactContext` utility, so ideally you would not add new functionality to it. Instead you could create a new React context to pass you `KibanaEnvironment` state.

```tsx
export const myEnvContext = React.createContext(/* ... */);
```

jennypavlova committed:
Merge branch 'main' into 172506-obsux-add-feedback-form-to-apm

jennypavlova committed:
Move KibanaEnvironmentContext to apm and infra

jennypavlova commented:
> In the long term we are planning to remove the `KibanaReactContext` utility, so ideally you would not add new functionality to it. Instead you could create a new React context to pass you `KibanaEnvironment` state.
> 
> ```tsx
> export const myEnvContext = React.createContext(/* ... */);
> ```

@vadimkibana Thank you for your review and for letting me know, I didn't know that. I moved the context inside of the plugins. And I reverted the changes inside the `KibanaReactContext`  (as suggested). It's sad that the utility will be removed - do you know if any alternative will be provided? I like the concept of having this shared Kibana context instead of moving it inside the plugins and duplicating part of the logic (Maybe a different shared package or something similar?)

crespocarlos commented on x-pack/plugins/apm/public/plugin.ts:
```diff
@@ -393,18 +403,26 @@ export class ApmPlugin implements Plugin<ApmPluginSetup, ApmPluginStart> {
-          pluginsStart: pluginsStart as ApmPluginStartDeps,
+          pluginsStart: {
+            ...pluginsStart,
+            isCloudEnv,
+            isServerlessEnv,
+            kibanaVersion: this.kibanaVersion,
```
I'm wondering whether these could go into a separate object named `kibanaEnv` or something like that and decouple them from `pluginsStart`. 

The only difference is that in the `renderApp` we would add another property to  `apmPluginContextValue`.

The same goes for what we currently have in the `infra` plugin.  

My concern is that we're adding things that are not related to plugin dependencies in the `pluginsStart` object

jennypavlova replied:
Thank you for the comments! IMO the Kibana environment data belongs to the `KibanaReactContext` as it's something related to the Kibana environment and it's not part of the plugin ( as it is the same for all plugins and it's something that we get as a dependency - same as the services and it should be shared ) But based on the [comment](https://github.com/elastic/kibana/pull/173758#pullrequestreview-1800192433) this context will be removed so I am not sure if we want to keep the concept of sharing a context (as package/utility) and reuse it. I don't know what will replace the `KibanaReactContext` but it can be a good place to have this environment-specific data independently there.  
So you are right that with the last changes is better to get them out of the `pluginsStart`. I will do that and I think for now we can keep them in the plugin context until we find a better solution.

jennypavlova replied:
So for both infra and apm now I am passing `kibanaEnvironment` object separately, can you please take a look again :)

crespocarlos commented on x-pack/plugins/infra/public/apps/common_providers.tsx:
```diff
@@ -70,11 +73,15 @@ export const CoreProviders: React.FC<CoreProvidersProps> = ({
     pluginStart
   );
 
+  const MyEnvContextForPluginProvider = useKibanaEnvironmentContextProvider(plugins);
```
suggestion: completely up to you
```suggestion
  const KibanaEnvContextForPluginProvider = useKibanaEnvironmentContextProvider(plugins);
```

jennypavlova replied:
Done, thanks ✅

crespocarlos reviewed (commented):
Hey Jenny. I took another look at it, and I'm just concerned about the stuff we're adding to PluginStartDeps, both in infra (possibly tech debt) and now in APM.  

If we compare with the rest of the code base, the start deps are supposed to be used for dependencies with other plugins and these flags don't quite fit in this scenario. Might just be me worrying too much, so I'd like to hear your opinion on this.

jennypavlova committed:
Merge branch 'main' of github.com:elastic/kibana into 172506-obsux-add-feedback-form-to-apm

jennypavlova committed:
Move kibana env from start deps in APM

jennypavlova committed:
Move kibana env from start deps in Infra

jennypavlova committed:
Merge branch 'main' into 172506-obsux-add-feedback-form-to-apm

jennypavlova committed:
Make KibanaEnv optional in common providers

jennypavlova committed:
Merge branch 'main' of github.com:elastic/kibana into 172506-obsux-add-feedback-form-to-apm

jennypavlova committed:
Make KibanaEnv optional in common providers and fix logs page

jennypavlova committed:
Merge branch 'main' into 172506-obsux-add-feedback-form-to-apm

jennypavlova committed:
Merge branch 'main' into 172506-obsux-add-feedback-form-to-apm

crespocarlos reviewed (approved):
Looks great! Thanks for making the changes :)
